### ISSUE TITLE
Search all possible components I can deploy

### ISSUE THREAD

**Issue by jorgemoralespou** [redhat-developer/ocdev#194]
As a developer, I want to be able to list/search all possible components I can use with my application

```
$ ocd catalog search [postgres] [--type=database|java|whatever] 
The following items were found:
(postgresql) PostgreSQL
(postgresql-10.0) PostgreSQL-10.0
(crunchydb) CrunchyDB Clustered (PostgreSQL 9.5)
```
Search has to be able to filter. (Possibly create another issue for this).

Subtask for #154

**Comment by kadel** [comment]
related to #54

**Comment by containscafeine** [comment]
I'll start looking into how the Service Catalog API can help us with this

**Comment by containscafeine** [comment]
@jorgemoralespou so I just tried on OpenShift versions v3.7.1 and up on minishift, but service catalog is not running by default, unless started otherwise using `minishift start --service-catalog`.
Since this is not the default behavior, imo it's not safe to only rely on the service catalog API to provide us the template names which the user can use in `ocdev component <type>`.

We also need to list the imagestreams in openshift namespace, in addition to templates being provided by the service catalog.

I think there are 2 ways we can do this,
- pause service catalog for now
- concatenate output of `oc get is -n openshift` and `oc get template -n openshift` and search through this

Or,
- assume that the service broker is enabled on the cluster (not the default behavior), and get templates from service catalag
- concatenate that output with `oc get is -n openshift`

WDYT?

**Comment by kadel** [comment]
@containscafeine why do you want to list all ImageStreams in the openshfit namespace (`oc get is -n openshift`)?

**Comment by jorgemoralespou** [comment]
@marekjelen @gshipley, any opinion on this?

@sspeiche is the catalog not going to be enabled by default? I thought we agreed to have it by default already.

**Comment by containscafeine** [comment]
> why do you want to list all ImageStreams in the openshfit namespace (oc get is -n openshift)?

How do we get the builder images that we currently support with ocdev?

**Comment by kadel** [comment]
> How do we get the builder images that we currently support with ocdev?

you have to add an additional filter to it, you will get more than just s2i builder images if you list all `is` in `openshift` namespace.

**Comment by marekjelen** [comment]
@jorgemoralespou in general I think we should target service catalog, it's default in Ansible installer anyway ... the question is - in case it's not gonna be default in `oc cluster up` / `minishift` - is the audience that will be using these two methods big enough to validate not using the nice stuff provided by the service catalog or would it better to ask for the switch during startup?

Just a spontaneous idea - should `od` have something like `od start` that would download `minishift` and would start ti with the correct switches? Yes, it's another layer of abstraction, but would simplify the setup for the user by hiding another dependency.

**Comment by containscafeine** [comment]
will start once #241 is merged

**Comment by sspeiche** [comment]
The service-catalog is not on by default. It is unclear when it will be. The direction is looking like more and more things will be enabled, like via concepts like alm in tectonic, though not 100% if catalog fits in as something that gets.

Having said that, I believe that in the next 3 months most hosted clusters would have it enabled. So for this effort, it would be good if ocdev could probe to the cluster to see if enabled and proceed accordingly

**Comment by jorgemoralespou** [comment]
ocdev as a tool for developers should be meant for clusters with the catalog enabled, where you can just use/deploy technology on a self service manner as a developer. 

I would maybe then:

- a) have that as pre-requisite
- b) delay to a posterior face where we want create an API on OpenShift that gives you the information independent of the catalog (but I guess we'll be on the same spot) 
- c) make the client intelligent on the capabilities of the cluster and use the appropriate apis when needed.

I guess a) for now and c) in the future
